### PR TITLE
Bug 1040548 - Move state (input sequence) out of IME engines.

### PR DESCRIPTION
**Pull request by RudyLu**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5c393416bbe400dd3796f313a2bd3a6ca25c58cf

File: apps/keyboard/js/keyboard/target_handlers.js [hunk 1]
```diff
@@ -62,16 +62,10 @@ DefaultTargetHandler.prototype.commit = function() {
    * maintained out of latin.js.
```
Comment (timdream) [inline apps/keyboard/js/keyboard/target_handlers.js]: Update the comment?

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=476a6f3feac8d72d9ac1815e8cb153bc3b12641e

File: apps/keyboard/js/keyboard/target_handlers.js (hunk 1 above)
Comment (RudyLu) [inline apps/keyboard/js/keyboard/target_handlers.js]: Thanks, comment updated.

File: apps/keyboard/js/keyboard/target_handlers.js [hunk 2]
```diff
@@ -54,24 +54,15 @@ DefaultTargetHandler.prototype.commit = function() {
   var engine = this.app.inputMethodManager.currentIMEngine;
 
   /*
-   * XXX: A hack to send both keycode and uppercase keycode to latin IME,
-   * since latin IME would maintain a promise queue for each key, and
-   * send correct keycode based on the current capitalization state.
-   * See bug 1013570 and bug 987809 for details.
-   * This hack should be removed and the state/input queue should be
-   * maintained out of latin.js.
+   * Now there is no inpu queue in latin.js, so we have to return promise here,
```
Comment (timdream) [inline apps/keyboard/js/keyboard/target_handlers.js]: `Return promise here, and ...`

Don't mentioned the removed behavior.

File: apps/keyboard/js/imes/latin/latin.js [hunk 3]
```diff
@@ -452,57 +444,47 @@
    *
    * Update the capitalization state, if we're capitalizing
    */
-  function click(keyCode, upperKeyCode, repeat) {
-    // Wait for the previous keys have been resolved and then handle the next
-    // key.
-
-    var nextKeyPromise = inputSequencePromise.then(function() {
-      keyCode = keyboard.isCapitalized() && upperKeyCode ? upperKeyCode :
```
Comment (timdream) [inline apps/keyboard/js/imes/latin/latin.js]: Should we remove `isCapitalized()` from the `InputMethodGlue` as well?

File: apps/keyboard/js/keyboard/target_handlers.js [hunk 4]
```diff
@@ -54,24 +54,15 @@ DefaultTargetHandler.prototype.commit = function() {
...
+   * and rely on the action queue in TargetHandlerManager to make sure the key
+   * code is sent with the right capitalization state.
    */
   var promise;
```
Comment (timdream) [inline apps/keyboard/js/keyboard/target_handlers.js]: Move this to line 65?

File: apps/keyboard/js/keyboard/target_handlers_manager.js [hunk 5]
```diff
@@ -107,7 +107,7 @@ TargetHandlersManager.prototype._callTargetAction = function(action,
   }
 
   this.promiseQueue = this.promiseQueue.then(function() {
-    handler[action](press);
+    return handler[action](press);
```
Comment (timdream) [inline apps/keyboard/js/keyboard/target_handlers_manager.js]: Ouch. Could you do the 319MB performance test again? It looks like we do not wait for API round trip in the previous patch.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f3d21f0df437981bfa3e69fd0a97bb25e3c1ef6f

File: apps/keyboard/js/imes/latin/latin.js (hunk 3 above)
Comment (RudyLu) [inline apps/keyboard/js/imes/latin/latin.js]: Done, thanks.

File: apps/keyboard/js/keyboard/target_handlers_manager.js (hunk 5 above)
Comment (RudyLu) [inline apps/keyboard/js/keyboard/target_handlers_manager.js]: Yes, just check this again on Flame with 319MB and the performance should be fine.

**Pull request merged**